compressors/huffman_coder.py

```diff
@@ -0,0 +1,279 @@
+from ast import List
+from concurrent.futures import BrokenExecutor
+from ctypes import Union
+from dataclasses import dataclass
+from typing import Any
+import heapq
+from functools import total_ordering
+from core.data_stream import BitsDataStream, DataStream
+from core.prob_dist import ProbabilityDist
+from core.data_compressor import DataCompressor
+from core.data_transformer import (
+    BitstringToBitsTransformer,
+    CascadeTransformer,
+    BitsParserTransformer,
+    LookupTableTransformer,
+)
+
+
+@dataclass
+@total_ordering  # decorator which adds other compare ops give one
+class HuffmanNode:
+    left_child: Any = None
+    right_child: Any = None
+    id: Any = None
+    code: str = ""
+    prob: float = None
+
+    @property
+    def is_leaf_node(self):
+        return (self.left_child is None) and (self.right_child is None)
+
+    def get_encoding_table(self):
+        """
+        parse the tree and get the encoding table
+        """
+
+        if self.is_leaf_node:
+            return {self.id: self.code}
+
+        encoding_table = dict()
+        if self.left_child is not None:
+            self.left_child.code = self.code + "0"
+            left_table_dict = self.left_child.get_encoding_table()
+            encoding_table.update(left_table_dict)
+
+        if self.right_child is not None:
+            self.right_child.code = self.code + "1"
+            right_table_dict = self.right_child.get_encoding_table()
+            encoding_table.update(right_table_dict)
+
+        return encoding_table
+
+    def __le__(self, other):
+        """
+        Define a comparison operator, so that we can use this while comparing nodes
+        # NOTE: we only need to define one compare op, as others can be implemented using the
+        decorator @total_ordering
+        """
+        return self.prob <= other.prob
+
+    def _get_lines(self):
```

Comment:
@shubhamchandak94: this is the code I added ... please review!
I guess we should move to a generic Prefix code util instead of just huffman